# **ESTUDO DE CASO PARA MATÉRIA DE MINERAÇÃO DE DADOS**

In [3]:
import pandas as pd

df = pd.read_csv(r'C:\Users\João Pedro\Documents\UFG\MD\estudo_caso\data\UCMF_raw.csv')
df.head()

ID  Peso  Altura   IMC Atendimento        DN  IDADE Convenio   PULSOS  \
0   1   5.0      51  19.0    11/05/06  30/03/06   0.12       GS  Normais   
1   2   3.5      50  14.0    25/05/05  19/05/05   0.02       GS  Normais   
2   3   0.0       0   NaN    12/06/01  08/05/05  -4.05     SULA  Normais   
3   4   8.1      65  19.0    15/10/09  21/04/09   0.50      NaN  Normais   
4   7  40.0     151  18.0    14/01/08  14/08/95  12.89     SAME  Normais   

   PA SISTOLICA  ...            PPA NORMAL X ANORMAL      B2      SOPRO   FC  \
0           NaN  ...  Não Calculado          Anormal  Normal  Sistólico  112   
1           NaN  ...  Não Calculado          Anormal  Normal    ausente  128   
2           NaN  ...  Não Calculado          Anormal  Normal  Sistólico   88   
3           NaN  ...  Não Calculado          Anormal  Normal    ausente   92   
4           NaN  ...  Não Calculado          Anormal  Normal    ausente   96   

            HDA 1 HDA2 SEXO                      MOTIVO1  \
0      Palpitacao  NaN    M  6 - Suspeita de cardiopatia   
1        Dispneia  NaN    M  6 - Suspeita de cardiopatia   
2   Assintomático  NaN    M                 2 - Check-up   
3   Assintomático  NaN    M     5 - Parecer cardiológico   
4  Dor precordial  NaN    M     5 - Parecer cardiológico   

                               MOTIVO2  
0  6 - Palpitação/taquicardia/arritmia  
1                         6 - Dispnéia  
2                                  NaN  
3                                  NaN  
4                                  NaN  

[5 rows x 21 columns]

Tamanho do conjunto de dados é:

In [4]:
len(df)

12873

## Retirando a coluna *ID*

In [5]:
df_copy = df.copy()
df_copy.drop(columns=['ID'], inplace=True)
df_copy.head()

Peso  Altura   IMC Atendimento        DN  IDADE Convenio   PULSOS  \
0   5.0      51  19.0    11/05/06  30/03/06   0.12       GS  Normais   
1   3.5      50  14.0    25/05/05  19/05/05   0.02       GS  Normais   
2   0.0       0   NaN    12/06/01  08/05/05  -4.05     SULA  Normais   
3   8.1      65  19.0    15/10/09  21/04/09   0.50      NaN  Normais   
4  40.0     151  18.0    14/01/08  14/08/95  12.89     SAME  Normais   

   PA SISTOLICA  PA DIASTOLICA            PPA NORMAL X ANORMAL      B2  \
0           NaN            NaN  Não Calculado          Anormal  Normal   
1           NaN            NaN  Não Calculado          Anormal  Normal   
2           NaN            NaN  Não Calculado          Anormal  Normal   
3           NaN            NaN  Não Calculado          Anormal  Normal   
4           NaN            NaN  Não Calculado          Anormal  Normal   

       SOPRO   FC           HDA 1 HDA2 SEXO                      MOTIVO1  \
0  Sistólico  112      Palpitacao  NaN    M  6 - Suspeita de cardiopatia   
1    ausente  128        Dispneia  NaN    M  6 - Suspeita de cardiopatia   
2  Sistólico   88   Assintomático  NaN    M                 2 - Check-up   
3    ausente   92   Assintomático  NaN    M     5 - Parecer cardiológico   
4    ausente   96  Dor precordial  NaN    M     5 - Parecer cardiológico   

                               MOTIVO2  
0  6 - Palpitação/taquicardia/arritmia  
1                         6 - Dispnéia  
2                                  NaN  
3                                  NaN  
4                                  NaN

## Análise de dados faltantes nas colunas

### Peso

In [6]:
print("Informações gerais sobre a coluna 'Peso':")
print(df_copy['Peso'].describe())

print("\n"+"-"*20)

print("\nNúmero de valores faltantes na coluna 'Peso':")
print(df_copy['Peso'].isna().sum())

Informações gerais sobre a coluna 'Peso':
count    12555.000000
mean        16.614417
std         16.979814
min        -40.000000
25%          3.400000
50%         12.800000
75%         24.000000
max        157.000000
Name: Peso, dtype: float64

--------------------

Número de valores faltantes na coluna 'Peso':
318


Temos por volta de 2.5% dos dados dessa coluna faltante, não é muito preocupante. Preocupante é peso negativo e outros dados estranhos.

### Altura

In [7]:
print("Informações gerais sobre a coluna 'Altura':")
print(df_copy['Altura'].describe())

print("\n"+"-"*20)

print("\nNúmero de valores faltantes na coluna 'Altura':")
print(df_copy['Altura'].isna().sum())

Informações gerais sobre a coluna 'Altura':
count    12873.000000
mean        66.111474
std         56.027934
min          0.000000
25%          0.000000
50%         69.000000
75%        112.000000
max        198.000000
Name: Altura, dtype: float64

--------------------

Número de valores faltantes na coluna 'Altura':
0


Não temos dados faltantes, isso é muito bom. Nem alturas negativas e dados estranhos nessa primeira análise.

### IMC

In [8]:
print("Informações gerais sobre a coluna 'IMC':")
print(df_copy['IMC'].describe())

print("\n"+"-"*20)

print("\nNúmero de valores faltantes na coluna 'IMC':")
print(df_copy['IMC'].isna().sum())

Informações gerais sobre a coluna 'IMC':
count    8146.000000
mean       17.455315
std        15.008016
min         0.000000
25%        15.000000
50%        17.000000
75%        19.000000
max       848.000000
Name: IMC, dtype: float64

--------------------

Número de valores faltantes na coluna 'IMC':
4727


Nessa coluna, possuímos 36.72% dos dados faltantes, isso já começa a ficar preocupante.

### Atendimento

In [9]:
print("\nNúmero de valores faltantes na coluna 'Atendimento':")
print(df_copy['Atendimento'].isna().sum())


Número de valores faltantes na coluna 'Atendimento':
983


Nessa coluna, temos cerca de 7.6% de dados faltantes. Não vejo isso como problema, dado que são poucos dados faltantes e não vejo muita importância para essa coluna (data da visita do paciente).

### DN (Data de Nascimento)

In [10]:
print("\nNúmero de valores faltantes na coluna 'DN':")
print(df_copy['DN'].isna().sum())


Número de valores faltantes na coluna 'DN':
1376


Nessa coluna, temos cerca de 10.7% de dados faltantes. Não vejo isso como problema, dado que são poucos dados faltantes e não vejo muita importância para essa coluna, já que possuímos uma coluna de idade (um pouco suspeita pela qualidade dos dados) e não sabemos quando esse conjunto de dados foi coletado, para fazer o cálculo das idades.

### Idade

In [11]:
print("Informações gerais sobre a coluna 'IDADE':")
print(df_copy['IDADE'].describe())

print("\n"+"-"*20)

print("\nNúmero de valores faltantes na coluna 'IDADE':")
print(df_copy['IDADE'].isna().sum())

Informações gerais sobre a coluna 'IDADE':
count    11377.000000
mean         4.856947
std          8.829742
min       -113.180000
25%          0.930000
50%          3.960000
75%          8.590000
max         71.810000
Name: IDADE, dtype: float64

--------------------

Número de valores faltantes na coluna 'IDADE':
1496


Temos cerca 11.5% de dados faltantes. Porém esse não é o maior problema, esses dados estão muito confusos e as informações que tenho sobre o conjunto de dados não colaboram para entendê-los. Temos uma idade de -113.18, de 0.01, etc. Como posso interpretar isso?

### Convênio

In [12]:
print("\nNúmero de valores faltantes na coluna 'Convenio':")
print(df_copy['Convenio'].isna().sum())


Número de valores faltantes na coluna 'Convenio':
4108


Temos cerca de 32% de dados faltantes nessa coluna. No entanto, isso não é preocupantes, pois ainda não vejo importância dessa coluna para o nosso objetivo, provavelmente vou retirá-la.

### Pulsos

In [13]:
print("\nNúmero de valores faltantes na coluna 'PULSOS':")
print(df_copy['PULSOS'].isna().sum())


Número de valores faltantes na coluna 'PULSOS':
1194


Temos cerca de 9% de dados faltantes.

### PA Sistólica e Diastólica

In [14]:
print("\nNúmero de valores faltantes na coluna 'PA SISTOLICA':")
print(df_copy['PA SISTOLICA'].isna().sum())

print("\n"+"-"*20)

print("\nNúmero de valores faltantes na coluna 'PA DIASTOLICA':")
print(df_copy['PA DIASTOLICA'].isna().sum())


Número de valores faltantes na coluna 'PA SISTOLICA':
7730

--------------------

Número de valores faltantes na coluna 'PA DIASTOLICA':
7740


Temos cerca de 60% de dados faltantes.

### PPA

In [15]:
print("\nNúmero de valores faltantes na coluna 'PPA':")
print(df_copy['PPA'].isna().sum())



Número de valores faltantes na coluna 'PPA':
217


In [16]:
df_copy['PPA'].value_counts()

PPA
Não Calculado          9081
Normal                 1790
#VALUE!                1496
HAS-2 PAS                73
Pre-Hipertensão PAD      73
Pre-Hipertensão PAS      48
HAS-1 PAS                48
HAS-1 PAD                26
HAS-2 PAD                21
Name: count, dtype: int64

Temos cerca de 1.7% de dados faltantes. No entanto, o problema é a quantiade de valores não calculados.

### Normal X Anormal

In [17]:
print("\nNúmero de valores faltantes na coluna 'NORMAL X ANORMAL':")
print(df_copy['NORMAL X ANORMAL'].isna().sum())


Número de valores faltantes na coluna 'NORMAL X ANORMAL':
1168


In [18]:
df_copy['NORMAL X ANORMAL'].value_counts()

NORMAL X ANORMAL
Normal     6744
Anormal    4959
anormal       1
Normais       1
Name: count, dtype: int64

Temos cerca de 9% de dados faltantes nessa coluna. No entanto, essa é a nossa coluna alvo, é a nossa label, o que estamos querendo prever. Então, as linhas que não possuem essa *label* vão ter que ser excluídas do nosso conjunto de dados.

### B2

In [19]:
print("\nNúmero de valores faltantes na coluna 'B2':")
print(df_copy['B2'].isna().sum())


Número de valores faltantes na coluna 'B2':
1178


In [20]:
df_copy['B2'].value_counts()

B2
Normal           11098
Hiperfonética      295
Desdob fixo        149
Única               77
Outro               76
Name: count, dtype: int64

Temos cerca de 9% de dados faltantes nessa coluna.

### Sopro

In [21]:
print("\nNúmero de valores faltantes na coluna 'SOPRO':")
print(df_copy['SOPRO'].isna().sum())


Número de valores faltantes na coluna 'SOPRO':
1167


In [22]:
df_copy['SOPRO'].value_counts()

SOPRO
ausente                   7275
Sistólico                 3648
sistólico                  723
contínuo                    28
Contínuo                    20
diastólico                   9
Sistolico e diastólico       3
Name: count, dtype: int64

Temos cerca de 9% de dados faltantes.

### Frequência Cardíaca

In [23]:
print("\nNúmero de valores faltantes na coluna 'FC':")
print(df_copy['FC'].isna().sum())


Número de valores faltantes na coluna 'FC':
1868


In [24]:
df_copy['FC'].describe()

count     11005
unique      118
top         100
freq       1509
Name: FC, dtype: object

Temos cerca de 14.5% de dados faltantes

### Histórico da doença

In [25]:
print("\nNúmero de valores faltantes na coluna 'HDA 1':")
print(df_copy['HDA 1'].isna().sum())

print("\nNúmero de valores faltantes na coluna 'HDA2':")
print(df_copy['HDA2'].isna().sum())


Número de valores faltantes na coluna 'HDA 1':
4270

Número de valores faltantes na coluna 'HDA2':
12466


Temos cerca de 33% dados faltantes para 'HDA 1' e quase 97% para 'HDA2'. Depois, vou juntá-las em uma só coluna, que fala sobre doença.

### Sexo

In [26]:
print("\nNúmero de valores faltantes na coluna 'SEXO':")
print(df_copy['SEXO'].isna().sum())


Número de valores faltantes na coluna 'SEXO':
4


In [27]:
df_copy['SEXO'].value_counts()

SEXO
M                6065
F                4477
Indeterminado    1417
Masculino         584
Feminino          247
masculino          79
Name: count, dtype: int64

Temos 0.03% de dados faltantes nessa coluna. Uma possibilidade será excluir esses dados, haja visto que não influenciarão muito.

### Motivo para ser encaminhado para clínica de cardiologia

In [28]:
print("\nNúmero de valores faltantes na coluna 'MOTIVO1':")
print(df_copy['MOTIVO1'].isna().sum())

print("\nNúmero de valores faltantes na coluna 'MOTIVO2':")
print(df_copy['MOTIVO2'].isna().sum())


Número de valores faltantes na coluna 'MOTIVO1':
1055

Número de valores faltantes na coluna 'MOTIVO2':
3496


In [29]:
print(df_copy['MOTIVO1'].value_counts())

print('-----------------------------')

print(df_copy['MOTIVO2'].value_counts())

MOTIVO1
5 - Parecer cardiológico           5352
6 - Suspeita de cardiopatia        4285
1 - Cardiopatia já estabelecida    1158
2 - Check-up                        701
7 - Outro                           322
Name: count, dtype: int64
-----------------------------
MOTIVO2
5 - Cirurgia                           2940
6 - Sopro                              2349
1 - Cardiopatia congenica              1041
Outro                                   824
5 - Atividade física                    652
6 - Palpitação/taquicardia/arritmia     415
6 - Dor precordial                      392
6 - HAS/dislipidemia/obesidade          240
6 - Dispnéia                            226
6 - Cianose                             145
1 - Cardiopatia adquirida                92
6 - Cardiopatia na familia               35
5 - Uso de cisaprida                      9
6 - Cansaço                               7
6 - Alterações de pulso/perfusão          5
6 - Cianose e dispnéia                    5
Name: count, dtype: int6

Temos cerca de 8.2% de dados faltantes para 'MOTIVO1' e 27% para 'MOTIVO2'. Uma opção pode ser juntar essas duas variáveis em uma única variável.

## Análise da relação dos atributos com as labels

In [30]:
target_col = 'NORMAL X ANORMAL'
mapeamento_alvo = {
    'Anormal': 'Anormal',
    'anormal': 'Anormal',
    'Normal': 'Normal',
    'Normais': 'Normal'
}
df[target_col] = df[target_col].map(mapeamento_alvo)
df[target_col].value_counts()

NORMAL X ANORMAL
Normal     6745
Anormal    4960
Name: count, dtype: int64

In [31]:
cat_cols = [col for col in df.select_dtypes(include='object').columns 
            if col != target_col and df[col].nunique() < 100]

print(f"Analisando as seguintes variáveis: {cat_cols}\n")


for col in cat_cols:
    print(f"--- Relação: {col} vs {target_col} ---")
    
    crosstab = pd.crosstab(df[col], df[target_col], dropna=False)
    
    crosstab['Total'] = crosstab.sum(axis=1)
    
    if 'Anormal' in crosstab.columns:
        crosstab['% Anormal'] = (crosstab['Anormal'] / crosstab['Total'] * 100).round(1)
    
    print(crosstab)
    print("\n" + "="*50 + "\n")

Analisando as seguintes variáveis: ['PULSOS', 'PPA', 'B2', 'SOPRO', 'HDA 1', 'HDA2', 'SEXO', 'MOTIVO1', 'MOTIVO2']

--- Relação: PULSOS vs NORMAL X ANORMAL ---
NORMAL X ANORMAL     Anormal  Normal   NaN  Total  % Anormal
PULSOS                                                      
AMPLOS                     1       0     0      1      100.0
Amplos                    45       7     0     52       86.5
Diminuídos                16       2     0     18       88.9
Femorais diminuidos       36       3     0     39       92.3
NORMAIS                    2       0     0      2      100.0
Normais                 4811    6698    22  11531       41.7
Outro                     26      10     0     36       72.2
NaN                       23      25  1146   1194        1.9


--- Relação: PPA vs NORMAL X ANORMAL ---
NORMAL X ANORMAL     Anormal  Normal  NaN  Total  % Anormal
PPA                                                        
#VALUE!                  286     434  776   1496       19.1
HAS-1 P